app/models/course/discussion/topic.rb

```diff
@@ -19,7 +19,9 @@ def subscribed_by?(user)
   # @param [User] user The user who needs to subscribe to this topic
   def ensure_subscribed_by(user)
     return true if self.subscribed_by?(user)
-    subscriptions.build(user: user).save!
+    ActiveRecord::Base.transaction do
+      subscriptions.build(user: user).save!
```

Comment:
any errors caused by a statement in a transaction will require the entire transaction to be rolled back.

here we expect the error -- the duplicate key cannot be guarded against. so the give it a transaction to error, we rollback, and the rest of the program continues.